Add preliminary HLSL support

jeremyong opened the pull request:
- Adds a new language mode to monaco, extending the base C++ layer with HLSL intrinsics and types
- Adds a new `HLSLCompiler` class
- Adds a sample pixel shader

The compiler used to test this locally is the [DirectX Shader Compiler](https://github.com/microsoft/DirectXShaderCompiler) (aka DXC), which needs to also be added to the [infra](https://github.com/compiler-explorer/infra) project.

Some guidance is needed before this PR can be merged:

1. While DXC can run on Linux, there are no binaries available so this
   compiler must be built. Are there examples that show how we should do
   this as part of the infra CI/CD? Should we build and host it
   separately instead? The build process for DXC on Linux is relatively
   straightforward and documented
   [here](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DxcOnUnix.rst).
2. The example code doesn't compile unless the user also supplies
   additional compiler flags `-T ps_6_6 -E PSMain`. Is there a way to
   load these flags conditionally only if the sample is loaded?
3. Technically, DXC emits DXIL IR (based on LLVM IR) and I am wondering
   if it's possible to extend an existing LLVM backend. In addition, the
   `-spirv` compiler flag could be emitted to target the SPIR-V backend
   instead, so I'm curious if there is a good way to express the target
   backend.

Sample DXIL output:
![image](https://user-images.githubusercontent.com/250149/182011352-a23526ce-2806-491d-a65e-abd7e0c06a40.png)

Sample SPIR-V output:
![image](https://user-images.githubusercontent.com/250149/182011343-22541ae3-ec54-421a-ad1d-672d695a27f4.png)

jeremyong commented:
This PR partially addresses issue #2331 by adding HLSL support, but for GLSL, the [`glslang`](https://github.com/KhronosGroup/glslang) compiler is needed.

RubenRBS commented on lib/compilers/hlsl.js:
```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2021, Compiler Explorer Authors
```
Header date is wrong

jeremyong-az replied:
Thanks this was taken from a different file. Random question: Are these copyright headings needed for this project in general?

RubenRBS commented on static/modes/hlsl-mode.ts:
```diff
@@ -0,0 +1,322 @@
+// Copyright (c) 2017, Compiler Explorer Authors
```
Date header is wrong

RubenRBS commented on static/modes/hlsl-mode.ts:
```diff
@@ -0,0 +1,322 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+'use strict';
```
No need for strict in ts code, we need to remove it from the rest of the ts mode files

RubenRBS commented on lib/languages.ts:
```diff
@@ -293,6 +293,16 @@ const definitions: Record<LanguageKey, LanguageDefinition> = {
+        monaco: 'hlsl',
+        extensions: ['.hlsl', '.hlsli'],
+        alias: [],
+        logoUrl: 'dxc.png',
```
Doesn't HLSL have a log that is not the one from the dxc? I think it would make more sense like that

jeremyong-az replied:
HLSL doesn't have a logo to my knowledge, and it seems like a logo is mandatory at the moment. I meant to ask about this actually because this image is just a placeholder. Do you have suggestions?

RubenRBS replied:
Oh. I've fixed it in #3934, sorry about that. It might take a few days for it to land, but once it's there we should be ok just leaving this `null`

jeremyong replied:
Ah no worries, I'll look out for that PR then (also I realized I was commenting from the wrong github account, sorry about that, @jeremyong-az is indeed also me)

RubenRBS commented on etc/config/hlsl.defaults.properties:
```diff
@@ -0,0 +1,3 @@
+compilers=/usr/dxc-artifacts/bin/dxc
```
This code will only work for default users. Once we have a corresponding infra PR, we can see what we need to add into an eventual `hlsl.amazon.properties` file

mattgodbolt replied:
We can do this change once things are built ok!

jeremyong replied:
OK for now I've removed this file and we can add things back once binaries are available through s3

RubenRBS reviewed (approved):
Hi! Thanks for taking the time to implement this, we really appreciate it.

As for the questions:
> 1. While DXC can run on Linux, there are no binaries available so this
compiler must be built [...]

This should be done in the infra repository & the appropiate build-{type} one. There are some examples of PRs in those repos that can give you a hint into how to do that, but please do let us know if you have any questions, we're more than happy to help! If in the process you're able to also add the glsl compiler, that would be great, but don't sweat it if it's too much initial work. We can pick up from this initial implementation and add it later on

> 2. The example code doesn't compile unless the user also supplies
additional compiler flags [...]

For example codes, no, sadly there's no way to set the compiler options currently other than describing it in the source code itself. We might want to look into that, but that's for another time :)


> 3. Technically, DXC emits DXIL IR [...]

If I'm understaning this one right, yes! There should be. But I'm not the most knowledgable person about that so I'll let someonelse answer that.


Again, thanks so much for your time, and sorry about the nitpicking PR review comments!

jeremyong commented:
Thanks for the feedback @RubenRBS. I'll try to see if there's an infra PR I can use to base my testing on. Do you have suggestions for testing if I don't have a Linux machine on-hand at the moment, perhaps through Github Actions? (I'm working in a personal capacity to integrate this).

RubenRBS commented:
That's something that maybe @mattgodbolt or @partouf can better answer :)

partouf commented:
> Thanks for the feedback @RubenRBS. I'll try to see if there's an infra PR I can use to base my testing on. Do you have suggestions for testing if I don't have a Linux machine on-hand at the moment, perhaps through Github Actions? (I'm working in a personal capacity to integrate this).

Infra is the least important, we can fix that for you.
The important part is the building, and you should be able to do that with Docker on either Windows or Mac.

You can copy paste https://github.com/compiler-explorer/misc-builder/blob/main/build/build-beebasm.sh and replace the necessary commands to clone and build according to https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DxcOnUnix.rst
I would start with just trying to build version v1.7.2207

You can run Docker on Mac by following the instructions on https://github.com/compiler-explorer/misc-builder

For Windows I think (but not entirely sure) you'll need to run something like (in the docker commandline / powershell)
```
docker build -t miscbuilder .
docker run miscbuilder sh build-yourcompiler.sh 1.7.2207
```

jeremyong commented:
@partouf Thanks for the suggestions, I have now incorporated those suggestions and have a working Linux-built DXC here: https://github.com/compiler-explorer/misc-builder/pull/43

jeremyong commented:
Hey team, I've rebased this PR again and removed the placeholder logo now that a null language logo is permitted. Would still very much appreciate feedback on https://github.com/compiler-explorer/misc-builder/pull/43 to add DXC

llvm-beanz commented:
I just want to give a huge shout-out to this. I'm super excited to see this in progress. There's an ongoing effort to add HLSL support into clang (https://clang.llvm.org/docs/HLSLSupport.html), and having DXC support in Compiler Explorer will be massively useful.

RubenRBS commented:
Hi! Sorry for the delay. Things tend to move slowly this time of the year as most of us are afk :)

I'll try to get this merged today

mattgodbolt commented:
Thanks all; my own AFKness is the root of this: I hope to clear tomorrow evening for this stuff if nobody else beats me to it! Thanks for your patience.

mattgodbolt commented:
(Can you take a look at fixing the style violations, missing smeicolons etc. Thanks!)

jeremyong commented:
Hey sure thing will look into that today. Fyi the git hooks don't seem to work on Windows so I had to commit with no-verify which explains the style violations. Appreciate eyes on this!

mattgodbolt commented on examples/hlsl/default.hlsl:
```diff
@@ -0,0 +1,22 @@
+// The entry point and target profile are needed to compile this example:
+// -T ps_6_6 -E PSMain
```
This is cool for now. (We need a more general solution for the "we'd like to provide arguments but allow users to override them" type of thing. We've never found a good solution on our side.)

mattgodbolt replied:
thanks!

ojeda replied:
Yeah, it is similar to https://github.com/compiler-explorer/compiler-explorer/issues/3765.

mattgodbolt commented on examples/hlsl/default.hlsl:
```diff
@@ -0,0 +1,22 @@
+
+float4 PSMain(PSInput input) : SV_Target0
+{
+    return color.Sample(color_sampler, input.uv);
```
It's silly I know but all our examples "square" their inputs. Is that something that even makes sense in this context?

mattgodbolt commented on static/modes/hlsl-mode.ts:
```diff
@@ -0,0 +1,321 @@
+        'WaveMultiPrefixOr',
+        'WaveMultiPrefixXor',
+
+        // Raytracing intrinsics
```
Blimey....shaders have come a long way since my day hacking on Xbox games :D

mattgodbolt reviewed (approved):
LGTM! Thanks so much for this. Will work on the infra stuff and once the tests et al are green (and you've had a chance to reply to the minimal feedback I have), then we can merge!

jeremyong commented:
I believe I've addressed the style issues using `npm run lint` as well as the feedback items. Thanks for the PR review! (will need another run to make sure I did things correctly)

mattgodbolt commented:
This is great: will merge and get things ready for a launch! Are there particular versions of the compiler I should configure. I'll do a "daily trunk build" compiler; it's fairly easy to add more. I'll CC you on the PRs/changes to add the relevant configuration.

jeremyong commented:
> This is great: will merge and get things ready for a launch! Are there particular versions of the compiler I should configure. I'll do a "daily trunk build" compiler; it's fairly easy to add more. I'll CC you on the PRs/changes to add the relevant configuration.

Awesome! Trunk builds are great and I would suggest the two trailing releases [here](https://github.com/microsoft/DirectXShaderCompiler/tags) also (the officially tagged DXC releases).

mattgodbolt commented:
Will do! Thanks :)

mattgodbolt commented:
The various compilers have built. I haven't yet had time to get the amazon config done (I do this on the commute to work and sadly the 5G connection kept dropping out today). Hopefully at lunch I can do something more!

mattgodbolt commented:
This is deploying now - let us know how it looks!

jeremyong commented:
Just checked, awesome work thanks so much! Really impressed with how mature all the infrastructure is :)
